packages/ics721/src/ibc_packet_receive.rs

```diff
@@ -52,10 +57,27 @@ pub(crate) fn receive_ibc_packet(
     packet: IbcPacket,
 ) -> Result<IbcReceiveResponse, ContractError> {
     PO.error_if_paused(deps.storage)?;
-
     let data: NonFungibleTokenPacketData = from_json(&packet.data)?;
     data.validate()?;
 
+    // If there is an incoming proxy, let proxy validate the packet, in case it fails, we fail the transfer
+    // This proxy for example whitelist channels that can send to this contract:
+    // https://github.com/arkprotocol/cw721-proxy/tree/main/contracts/cw721-incoming-proxy
+    let incoming_proxy_msg = match INCOMING_PROXY.load(deps.storage).ok().flatten() {
+        Some(incoming_proxy) => {
+            let msg = to_json_binary(&ReceiverExecuteMsg::Ics721ReceivePacketMsg {
+                packet: packet.clone(),
+                data: data.clone(),
+            })?;
+            Some(WasmMsg::Execute {
+                contract_addr: incoming_proxy.to_string(),
+                msg,
+                funds: vec![],
+            })
+        }
+        None => None,
+    };
+
```

Comment:
Why is this not a query? feels like it should be a query.

Reply:
Not sure, what u mean, query is only `INCOMING_PROXY.load(deps.storage)`, while in query.rs there is `QueryMsg::IncomingProxy {} => to_json_binary(&INCOMING_PROXY.load(deps.storage)?),`.

But what I can do is moving remaining part to helper.rs like `get_incoming_proxy_msg`.

Reply:
moved to helper.rs: `get_incoming_proxy_msg()`